### PR TITLE
REQUEST FOR CODE REVIEW ON RUBY ON RAILS RE-FORMER PROJECT

### PR DESCRIPTION
**Pull request by Moreneecoder**
In this project, we created a rails application demonstrating our knowledge of Forms. We implemented a form that collects data to create new users and another that can be used to edit existing users' data. We also display error messages when invalid user data is submitted. Some of the files we worked on are:

* new.html.erb file - For handling the new user creation view
* edit.html.erb file - For handling updates of existing users.
* routes.rb - We used Rails `resources` keyword to handle our users' route and used the `only` attribute to limit actions to `new`, `create`, `edit` and `update`.
* user.rb - We validated the form fields to be present before accepting them and interacting with the users table using Active Records.

### CODE REVIEW

File: app/views/users/new.html.erb [hunk 1]
```diff
@@ -0,0 +1,19 @@
+  <% if @user.errors.any?%>
+    <h6><%= pluralize(@user.errors.count, "error") %> prohibited this post from being saved:</h6>
+    <ul>
+      <% @user.errors.full_messages.each do |msg| %>
+        <li><span style="color:red"><%= msg %></span></li>
+      <% end %>
+  </ul>
+  <%end%>
```
Comment (enaburekhan) [inline app/views/users/new.html.erb]: Move the logic to a helper. There should be no logic in view. Views are only to render information that becomes available to them from the corresponding controller and not for generating it.

**Review by enaburekhan — CHANGES_REQUESTED** [review]
# Status: Changes required:full_moon:
**Hi team!
Great job on this project:clap::clap:.**

**Major highlights**

- The PR and readme file is very descriptive.
- It's possible to easily create and edit a new user.
- Validations are present for all the fields.

However, we still need to make a few changes to get the project approved:muscle:.

**To improve**
- [ ] Add all the commented-out forms and leave them still commented out. Need a reminder? Check [odin](https://www.theodinproject.com/paths/full-stack-ruby-on-rails/courses/ruby-on-rails/lessons/forms) project.

- [ ] As highlighted below in the PR. Move the logic in new.html.erb under users in view to a helper. There should be absolutely NO logic in views. Views are only for displaying information that became available to it from its corresponding controller not generating it. Check this [article](http://codefol.io/posts/Where-Do-I-Put-My-Code/) and this [article](https://mixandgo.com/learn/the-beginners-guide-to-rails-helpers) if you need help on how to use a helper in rails. see [screen-shot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/section/no_logic_in_helpers_requirement.png) from pathwright

**Optional**
- [ ] Create a partial for user form and reused in edit and new under users in view. This will ensure DRY in the codebase.

Make the required changes and when you are ready, request another review.
Do have questions/comments? Kindly drop them :point_down: and ping me on slack @ Eric Enaburekhan to respond.
Happy hacking:blush:
Eric Enaburekhan

File: app/views/users/new.html.erb (hunk 1 above)
Comment (FdI96) [reply app/views/users/new.html.erb]: Excellent review. Thanks for the links. We will make the changes and request the PR again.

**Review by mke2111 — CHANGES_REQUESTED** [review]
# STATUS: CHANGES REQUIRED 🔴 ♻️ 

## Hey Happy Coders, 👋🏽 
**This is a great project you have going on, well-done lads and thanks for implementing the changes requested earlier on**. However, there are a few issues to fix to get approved.

### REQUIRED CHANGES: ⚠️ 

- [ ] The form tag in the new and edit HTML files are not properly commented out. You can enclose the entire block within this style of comment.
```
<%
=begin%>

     #Code to be commented out

<%
=end%>

```

- [ ] What the previous reviewer meant by moving the logic to helpers was that you create methods within the helpers directory that will contain your logic, and NOT create a partial name `helpers`. Use the given resources and understand how exactly to move the logic from all files in the views directory to the helpers directory.

- [ ] [OPTIONAL] For editing the existing user details, use a `form_for` tag as described here in [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/forms/user_form_edit.png) from [odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/forms)

### Kindly make all the requested changes and submit another CRR. 💯 

**Cheers 🍹 and Happy coding 💻** 

## Ping me on slack for a chit-chat 🆘  Roy Mukuye Or mention me in this PR

**Review by andresporras3423 — CHANGES_REQUESTED** [review]
### Status: Changes Required :recycle:
**Hello Morenikeji Bello & Federico Lopez Cechini** 
Good job with your new submission for this project :muscle:  :muscle:
It meets most of the requirements :clap: :clap: :clap:  
There are just a few details to improve before you can go ahead:

- [ ] It is expected you have two commented forms in the views/users/new.html.erb file, the first one using form_tag and the second one using form_with.
- [ ] Fix rubocop offenses.

### OPTIONAL

- [ ] It would be nice to add warning/error messages using flashcards to inform users about invalid data.
- [ ] It would be nice to add validation for valid email.
 

Please request another review once you are ready! :+1:

**Comment by FdI96** [comment]
About the first mandatory change, the last reviewer asks us to leave a commented-out form_tag and form_with structure but this is in contrast to the instructions from Odin Project which requires have an HTML form and a form_tag structure commented out. See these screenshots below for more information.

htmlform
htmlformtag

Then about the second mandatory change, I would like to share this screenshot from the GitHub pull request page, that demonstrates that all the checks have passed.

linterspassed

**Review by andresporras3423 — APPROVED** [review]
# Status: Approved :heavy_check_mark: :heavy_check_mark: :heavy_check_mark:
**Hello Morenikeji Bello & Federico Lopez Cechini** 
Amazing job :tada: :tada: :tada:  you have met all the requirements for this project
Now you are ready to continue with your next challenge :muscle: 
Don't forget to merge with the develop branch
 
Keep going like this  :clap: :clap:
 
Oscar Andrés Russi Porras

**Pull request merged**